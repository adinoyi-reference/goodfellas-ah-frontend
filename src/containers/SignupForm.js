import React, { Component } from 'react';
import { connect } from 'react-redux';
import { signup, clearSigninError } from '../actions/authActions';
import validateAuth from '../lib/validation';
import AuthInput from '../components/shared/AuthInput';
import Button from '../components/shared/Button';
import Loading from '../components/shared/Loading';
import '../styles/styles.scss';

const initialState = {
  // input fields
  firstname: '',
  lastname: '',
  email: '',
  password: '',
  confirmPassword: '',
  loading: false,
  touched: {
    firstname: false,
    lastname: false,
    email: false,
    password: false,
    confirmPassword: false,
  },
};

const fieldNames = ['firstname', 'lastname', 'email', 'password', 'confirmPassword'];

class SignupForm extends Component {
  state = initialState;

  componentDidMount = () => {
    const { clearSigninError: clearError } = this.props;
    clearError();
  }

  handleSubmit = (event) => {
    event.preventDefault();
    const { signup: signupUser, history } = this.props;
    const {
      firstname, lastname, email, password, confirmPassword,
    } = this.state;

    const fields = {
      firstname, lastname, email, password, confirmPassword,
    };

    // run validation rules for each field
    const validationError = validateAuth(fields, fieldNames);
    const changedTouchState = this.changeTouchState(fieldNames, true);
    this.setState({ touched: changedTouchState });

    const body = {
      firstname: firstname.trim(),
      lastname: lastname.trim(),
      email: email.trim(),
      password: password.trim(),
    };

    // processes user sign up when validation rules
    // pass for all fields
    if (!validationError.status) {
      this.setState({ loading: true });
      signupUser(body, () => history.push('/user/profile'));
      this.setState(initialState);
    }
  }

  handleChange = (event) => {
    this.setState({ [event.target.id]: event.target.value });
  }

  handleBlur = (field) => () => {
    const { touched } = this.state;
    this.setState({
      touched: { ...touched, [field]: true },
    });
  }

  changeTouchState = (fields, state) => {
    const touchState = {};
    fields.forEach((field) => {
      touchState[field] = state;
    });
    return touchState;
  }

  render() {
    const {
      firstname, lastname, email, password, confirmPassword, touched, loading,
    } = this.state;

    const { message } = this.props;
    const validationError = validateAuth({
      firstname, lastname, email, password, confirmPassword,
    }, fieldNames);

    return (
      <form className="auth-signup" onSubmit={this.handleSubmit}>
        <p className="error-field">{message}</p>
        <AuthInput
          error={validationError}
          value={firstname}
          touched={touched}
          handleChange={this.handleChange}
          handleBlur={this.handleBlur('firstname')}
          name="firstname"
          placeholder="First name"
        />

        <AuthInput
          error={validationError}
          value={lastname}
          touched={touched}
          handleChange={this.handleChange}
          handleBlur={this.handleBlur('lastname')}
          name="lastname"
          placeholder="Last name"
        />

        <AuthInput
          error={validationError}
          value={email}
          touched={touched}
          handleChange={this.handleChange}
          handleBlur={this.handleBlur('email')}
          name="email"
          placeholder="Email"
        />

        <AuthInput
          error={validationError}
          value={password}
          touched={touched}
          handleChange={this.handleChange}
          handleBlur={this.handleBlur('password')}
          name="password"
          type="password"
          placeholder="Password"
        />

        <AuthInput
          error={validationError}
          value={confirmPassword}
          touched={touched}
          handleChange={this.handleChange}
          handleBlur={this.handleBlur('confirmPassword')}
          name="confirmPassword"
          type="password"
          placeholder="Confirm Password"
        />
        {loading ? <button disabled="disabled" type="submit" className="auth-button loading"><Loading /></button> : <Button title="SIGN UP" className="auth-button" type="submit" />}
      </form>

    );
  }
}


const mapStateToProps = (state) => ({ message: state.auth.errorMessage });

export default connect(mapStateToProps, { signup, clearSigninError })(SignupForm);
