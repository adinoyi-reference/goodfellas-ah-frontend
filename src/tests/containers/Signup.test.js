import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import SignupForm from '../../containers/SignupForm';
import Root from '../../root';

let wrapped;

beforeEach(() => {
  wrapped = mount(
    <Root>
      <SignupForm />
    </Root>,
  );
});

afterEach(() => wrapped.unmount());

describe('Signup UI', () => {
  describe('render features', () => {
    test('container should render as expected', () => {
      const tree = toJson(wrapped);
      expect(tree).toMatchSnapshot();
    });

    it('has 5 input fields and a button', () => {
      expect(wrapped.find('input').length).toEqual(5);
      expect(wrapped.find('button').length).toEqual(1);
    });
  });

  describe('when typing into fields', () => {
    beforeEach(() => {
      wrapped.find('input').first().simulate('change', { target: { id: 'firstname', value: 'john' } });
      wrapped.find('input').at(1).simulate('change', { target: { id: 'lastname', value: 'doe' } });
      wrapped.find('input').at(2).simulate('change', { target: { id: 'email', value: 'johndoe@email.com' } });
      wrapped.find('input').at(3).simulate('change', { target: { id: 'password', value: 'password' } });
      wrapped.find('input').at(4).simulate('change', { target: { id: 'confirmPassword', value: 'password' } });
      wrapped.update();
    });

    it('firstname field has been filled', () => {
      expect(wrapped.find('input').first().prop('value')).toEqual('john');
    });

    it('lastname field has been filled', () => {
      expect(wrapped.find('input').at(1).prop('value')).toEqual('doe');
    });

    it('email field has been filled', () => {
      expect(wrapped.find('input').at(2).prop('value')).toEqual('johndoe@email.com');
    });

    it('password field has been filled', () => {
      expect(wrapped.find('input').at(3).prop('value')).toEqual('password');
    });

    it('confirmPassword field has been filled', () => {
      expect(wrapped.find('input').at(4).prop('value')).toEqual('password');
    });

    it('empties fields when form is submitted', () => {
      wrapped.find('form').simulate('submit');
      wrapped.update();

      expect(wrapped.find('input').first().prop('value')).toEqual('');
      expect(wrapped.find('input').at(1).prop('value')).toEqual('');
      expect(wrapped.find('input').at(2).prop('value')).toEqual('');
      expect(wrapped.find('input').at(3).prop('value')).toEqual('');
      expect(wrapped.find('input').at(4).prop('value')).toEqual('');
    });
  });

  describe('return empty field when form is submitted', () => {
    it('empties fields when form is submitted', () => {
      wrapped.find('form').simulate('submit');
      wrapped.update();

      expect(wrapped.find('input').first().prop('value')).toEqual('');
      expect(wrapped.find('input').at(1).prop('value')).toEqual('');
      expect(wrapped.find('input').at(2).prop('value')).toEqual('');
      expect(wrapped.find('input').at(3).prop('value')).toEqual('');
      expect(wrapped.find('input').at(4).prop('value')).toEqual('');
    });
  });
});
