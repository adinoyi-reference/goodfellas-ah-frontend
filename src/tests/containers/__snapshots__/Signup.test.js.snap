// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup UI render features container should render as expected 1`] = `
<_default>
  <Provider
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <Connect(SignupForm)>
      <SignupForm
        signup={[Function]}
      >
        <form
          onSubmit={[Function]}
        >
          <p
            className="error-field"
          />
          <AuthInput
            error={Object {}}
            handleChange={[Function]}
            name="firstname"
            placeholder="Firstname"
            value=""
          >
            <div
              className="auth-input-wrapper"
            >
              <label
                className="sr-only"
                htmlFor="firstname"
              >
                firstname
              </label>
              <input
                className="auth-field"
                id="firstname"
                onChange={[Function]}
                placeholder="Firstname"
                type="text"
                value=""
              />
              <div
                className="error-field"
              />
            </div>
          </AuthInput>
          <AuthInput
            error={Object {}}
            handleChange={[Function]}
            name="lastname"
            placeholder="Lastname"
            value=""
          >
            <div
              className="auth-input-wrapper"
            >
              <label
                className="sr-only"
                htmlFor="lastname"
              >
                lastname
              </label>
              <input
                className="auth-field"
                id="lastname"
                onChange={[Function]}
                placeholder="Lastname"
                type="text"
                value=""
              />
              <div
                className="error-field"
              />
            </div>
          </AuthInput>
          <AuthInput
            error={Object {}}
            handleChange={[Function]}
            name="email"
            placeholder="E-mail"
            type="email"
            value=""
          >
            <div
              className="auth-input-wrapper"
            >
              <label
                className="sr-only"
                htmlFor="email"
              >
                email
              </label>
              <input
                className="auth-field"
                id="email"
                onChange={[Function]}
                placeholder="E-mail"
                type="email"
                value=""
              />
              <div
                className="error-field"
              />
            </div>
          </AuthInput>
          <AuthInput
            error={Object {}}
            handleChange={[Function]}
            name="password"
            placeholder="Password"
            type="password"
            value=""
          >
            <div
              className="auth-input-wrapper"
            >
              <label
                className="sr-only"
                htmlFor="password"
              >
                password
              </label>
              <input
                className="auth-field"
                id="password"
                onChange={[Function]}
                placeholder="Password"
                type="password"
                value=""
              />
              <div
                className="error-field"
              />
            </div>
          </AuthInput>
          <AuthInput
            error={Object {}}
            handleChange={[Function]}
            name="confirmPassword"
            placeholder="Confirm Password"
            type="password"
            value=""
          >
            <div
              className="auth-input-wrapper"
            >
              <label
                className="sr-only"
                htmlFor="confirmPassword"
              >
                confirmPassword
              </label>
              <input
                className="auth-field"
                id="confirmPassword"
                onChange={[Function]}
                placeholder="Confirm Password"
                type="password"
                value=""
              />
              <div
                className="error-field"
              />
            </div>
          </AuthInput>
          <AuthButton
            name="SIGN UP"
          >
            <div>
              <button
                className="auth-button"
                type="submit"
              >
                SIGN UP
              </button>
            </div>
          </AuthButton>
        </form>
      </SignupForm>
    </Connect(SignupForm)>
  </Provider>
</_default>
`;
